### ISSUE TITLE
Saving a block with backslash in RichText wipes out all RichText data within the block

### ISSUE THREAD

**Issue by Alexius08** [WordPress/gutenberg#16508]
**Describe the bug**
Inputting backslashes into RichText components inside a PHP-rendered block results into loss of saved data inside all RichText components within the affected block.

**To reproduce**
1. Make a minimum variable block with at least two RichText input components and render it using PHP.
2. Use the block on a post, set the value of the RichText components to normal strings and save it.
3. Refresh the post editor for the post with the block and replace the contents of one of the RichText components with at least one backslash.
4. Preview to see the value of RichText-dependent attributes within the block reduced into empty strings. Save to overwrite all values with empty strings.

**Expected behavior**
Normal behavior is preserving all values of RichText-dependent attributes even when the backspace-filled string is inputted.

**Screenshots**
PHP-rendered block with two RichText inputs filled with normal values, as seen in the editor.
![PHP-rendering bug p1](https://user-images.githubusercontent.com/13504287/60973275-85e0d800-a35a-11e9-90fb-53cef86d5bc1.png)

PHP-rendered block with two RichText inputs filled with normal values, as seen in the preview.
![PHP-rendering bug p2](https://user-images.githubusercontent.com/13504287/60973428-dce6ad00-a35a-11e9-9d5a-5a43c4a80de6.png)

PHP-rendered block with two RichText inputs, one of them filled with only backslashes, as seen in the editor.
![PHP-rendering bug p3](https://user-images.githubusercontent.com/13504287/60973496-fa1b7b80-a35a-11e9-8d95-91a67943a3cc.png)

PHP-rendered block with two RichText inputs, one of them filled with only backslashes, as rendered in the preview. Saving the post at this state leads to removal of all data in the RichText inputs.
![PHP-rendering bug p4](https://user-images.githubusercontent.com/13504287/60973484-f2f46d80-a35a-11e9-93c7-fa4bf727c619.png)

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Browser: Chrome
 - Version: Wordpress 5.2.2, PHP 7.2.9

**Comment by ehti** [comment]
Able to reproduce this! Thanks for the report @Alexius08 

For easier testing, we can also use the Search block instead of creating a new block. Title in the search block is RichText and rendered in PHP.

1. Add search block to a post
2. Change the title from `Search` to `ABC\`. Save draft
3. Refresh the page
4. Title will be gone and it gets reset to just `Search`

(Backslashes are a problem).

Also ref https://wordpress.slack.com/archives/C02QB2JS7/p1576127699440400

**Comment by skorasaurus** [comment]
Hi, 

I'm still able to reproduce this as of Gutenberg 13.2.; note that this only occurs if the last character in the saved attributed is a backslash; 

for example, in the test case above by @ehti ; if you change the title in title block to "search for records and\or data" ; the attribute will be saved ;but if you finish it as ABC\ it will not.

**Comment by dmsnell** [comment]
@Mamaduka @youknowriad to be clear and get it out of the way, this isn't related to the recent `type: 'raw'` change is it?

**Comment by youknowriad** [comment]
@dmsnell I don't think so as that change is only applied to some blocks not all RichText blocks and this bug is about a custom block with two rich text elements.

**Comment by dmsnell** [comment]
thanks @youknowriad - I can imagine some plausible ways this could go wrong.

Can we get a copy of before/after `post_content`/HTML for these posts when the error is present?

**Comment by Mamaduka** [comment]
The search button text has a `type: 'string'`. This is what block HTML output looks like when I add a backslash at the end of the string.

Edit: I can only reproduce the issue when the string attribute is serialized in the block comment.

```html
<!-- wp:search {"label":"Search","buttonText":"Search \\u0022} /-->
```

## Screenshot
![CleanShot 2023-06-12 at 21 14 44](https://github.com/WordPress/gutenberg/assets/240569/fd8abd63-40be-46df-a854-adf729dae43c)

**Comment by Mamaduka** [comment]
I believe this regression was introduced in #6619.

**Comment by Mamaduka** [comment]
Hey, @dmsnell. Is this still on your radar?

Also cc-ing @sirreal, since he's been working on some similar areas recently.

**Comment by sirreal** [comment]
I took a quick look at this, here's my easy reproduction:

- Create a new post `/wp-admin/post-new.php`.
- Add a "More" block (`/more`).
- Replace the default text with the backslash character: `\`.

Save. The block is now broken. Checking the code editor reveals this:

```html
<!-- wp:more {"customText":"\\u0022} -->
<!--more \-->
<!-- /wp:more -->
```

The JSON is invalid. Reloading the page results in a block invalidation.

Here's a failing test:

```diff
diff --git a/packages/blocks/src/api/test/serializer.js b/packages/blocks/src/api/test/serializer.js
index 854e035b27d..eebe0fcc325 100644
--- a/packages/blocks/src/api/test/serializer.js
+++ b/packages/blocks/src/api/test/serializer.js
@@ -204,6 +204,14 @@ describe( 'block serializer', () => {
 				'{"a":"\\u0022 and \\u0022"}'
 			);
 		} );
+
+		it( 'should handle backslash and quote combinations', () => {
+			const orig = { bs: '\\', bsQuote: '\\"', bsQuoteBs: '\\"\\' };
+			expect( JSON.parse( serializeAttributes( orig ) ) ).toEqual( orig );
+			expect( serializeAttributes( orig ) ).toBe(
+				'{"bs":"\\\\","bsQuote":"\\\\\\u0022","bsQuoteBs":"\\\\\\u0022\\\\"}'
+			);
+		} );
 	} );
 
 	describe( 'getCommentDelimitedContent()', () => {
```

That may not be exactly the issue described here, but the stringified JSON looks like this:

```json
{"customText":"\\\\"}
```

Then this line matches the `\\"` and replaces it:

https://github.com/WordPress/gutenberg/blob/c436c1daa417a54cb4033fc42106b1a9b7bf1df8/packages/blocks/src/api/serializer.js#L296

Resulting in

```json
{"customText":"\\\\u0022}
```

I'm not sure whether #6619 introduced the issue, it's a likely candidate.

**Comment by sirreal** [comment]
This is tricky and requires Gutenberg and Core patches.

I've done some exploration and have draft PRs to address the issues:

- https://github.com/WordPress/wordpress-develop/pull/9558
- https://github.com/WordPress/gutenberg/pull/71291

I am concerned about [`wp_kses_stripslashes()` ](https://developer.wordpress.org/reference/functions/wp_kses_stripslashes/) that started this whole thing in the first place. If that's being applied to the JSON, I think it's going to break it again:

```php
$attrs = '<!-- wp:attributes {"key":"\\u0022\\\\"} /-->' ;
$json = '{"key":"\\u0022\\\\"}'; 
var_dump(
  $attrs,
  wp_kses_stripslashes($attrs),
  json_decode($json, true),
  json_decode(wp_kses_stripslashes($json)),
);
```

```
string(42) "<!-- wp:attributes {"key":"\u0022\\"} /-->"
string(41) "<!-- wp:attributes {"key":"\u0022\"} /-->"
array(1) {
  ["key"]=>
  string(2) ""\"
}
NULL
```

It's also appears to be missing the logic to not match `"` preceded by an escaped `\`.

**Comment by sirreal** [comment]
I created https://github.com/WordPress/wordpress-develop/pull/9559, but I'm unsure whether it's necessary.

**Comment by sirreal** [comment]
https://core.trac.wordpress.org/ticket/63917 is the Trac ticket for this, since it will require Core changes as well.

I'm proposing updated block attribute encoding to address the issue, ready for review.

- https://github.com/WordPress/wordpress-develop/pull/9558
- https://github.com/WordPress/gutenberg/pull/71291
